### PR TITLE
feat: create global notifications

### PR DESCRIPTION
**Pull request by Heba-WebDev**
[Linear Ticket BAC2-41](https://linear.app/team-bingo/issue/BAC2-41/create-notifications-for-all-users)

Description:

This pull request introduces a new feature for creating global notifications within the application. The feature allows super admins to send notifications to all users, facilitating effective communication regarding important updates, such as scheduled maintenance or new features.

Changes Made:

- [ x ] Implemented the endpoint POST /api/v1/notifications/global 
- [ x ] Added the createGlobalNotification method in the NotificationsService to manage the business logic for notification creation.
- [ x ]  Integrated input validation using DTO (Data Transfer Object) to ensure that the request payload is properly validated.
- [ x ] Added appropriate HTTP status codes and response messages for success and various error scenarios, including validation errors and unauthorized access.
- [ x ] Included unit tests for the NotificationsService to ensure the correctness of the notification creation logic.
- [ x ] Updated API documentation to reflect the new global notifications endpoint.

Testing Instructions:
- Login Super Admin:

 Use the /api/v1/auth/login endpoint with valid super admin credentials to obtain an access token.

- Create Notification:
Use the /api/v1/notifications/globals endpoint with the access token in the Authorization header and a valid message in the request body to create a global notification.

Success Response example:
`{
    "status": "success",
    "message": "Notification created successfully",
    "data": null
}`

### CODE REVIEW

File: src/modules/notifications/notifications.service.ts [hunk 1]
```diff
@@ -35,6 +37,29 @@ export class NotificationsService {
     private userRepository: Repository<User>
   ) {}
 
+  async createGlobalNotifications(dto: CreateNotificationForAllUsersDto) {
+    const users = await this.userRepository.find();
+    try {
```
Comment (Cobidex) [inline src/modules/notifications/notifications.service.ts:42]: there is no need for the try/cacth block. The interceptor will handle all non-specific errors appropriately

File: src/modules/notifications/notifications.service.ts [hunk 2]
```diff
@@ -35,6 +37,29 @@ export class NotificationsService {
...
+      });
+      await this.notificationRepository.save(notifications);
+      return {
+        status: 'success',
+        message: 'Notification created successfully',
+        data: null,
+      };
+    } catch (error) {
+      throw new CustomHttpException(
+        { message: 'An unexpected error occurred', error: error.message },
```
Comment (Cobidex) [inline src/modules/notifications/notifications.service.ts:57]: just a string containing the error message and the status code is expected by the CustomHttpException handler.  reference auth for illustration

File: src/modules/notifications/notifications.service.ts [hunk 3]
```diff
@@ -35,6 +37,29 @@ export class NotificationsService {
...
+          user,
+        });
+      });
+      await this.notificationRepository.save(notifications);
+      return {
+        status: 'success',
+        message: 'Notification created successfully',
+        data: null,
+      };
+    } catch (error) {
```
Comment (Cobidex) [inline src/modules/notifications/notifications.service.ts:55]: you need not handle server errors within the modules. we have an interceptor in place for this. You can remove the try / catch confidently

File: src/modules/notifications/tests/notifications.service.spec.ts [hunk 4]
```diff
@@ -49,13 +53,11 @@ describe('NotificationsService', () => {
         },
         {
           provide: getRepositoryToken(User),
-          useClass: Repository,
+          useValue: mockUserRepository, // Use the mockUserRepository here
```
Comment (Cobidex) [inline src/modules/notifications/tests/notifications.service.spec.ts:56]: remove dangling comment

**Review by Cobidex — COMMENTED** [review]
few comments. logic looks good to me at least for now. since its the same message for all users, why dont we just create a single super admin notification and have all users access the notification via another endpoint(almost like a publish/subscribe pattern). mapping through all users might be good for now but with a time complexity of O(n) it'll be less efficient as our user base grows to thousands.

File: src/modules/notifications/notifications.service.ts (hunk 1 above)
Comment (Heba-WebDev) [reply src/modules/notifications/notifications.service.ts]: Updated

File: src/modules/notifications/notifications.service.ts (hunk 2 above)
Comment (Heba-WebDev) [reply src/modules/notifications/notifications.service.ts]: Updated

File: src/modules/notifications/notifications.service.ts (hunk 3 above)
Comment (Heba-WebDev) [reply src/modules/notifications/notifications.service.ts]: Noted. Removed the try catch

File: src/modules/notifications/tests/notifications.service.spec.ts (hunk 4 above)
Comment (Heba-WebDev) [reply src/modules/notifications/tests/notifications.service.spec.ts]: Removed

**Comment by Heba-WebDev** [comment]
@Cobidex I added pagination and improved the users fetching by only selecting the id. This is still not optimized for medium to large applications. A better approach can be adding jobs/queue, but for that we'd need to add redis.

**Review by Cobidex — APPROVED** [review]

**Pull request merged**